### PR TITLE
AdaBelief bias correction

### PR DESCRIPTION
**Pull request by cossio**
Fix https://github.com/FluxML/Flux.jl/issues/1962

CC @darsnack @willtebbutt

### CODE REVIEW

File: src/optimise/optimisers.jl [hunk 1]
```diff
@@ -532,10 +532,26 @@ AdaBelief(η::Real, β::Tuple, state::IdDict) = AdaBelief(η, β, EPS, state)
...
+
+  #= st is a variance and can go to zero. This is in contrast to ADAM, which uses the
+  second moment which is usually far enough from zero. This is problematic, since st
+  can be slightly negative due to numerical error, and the square root below will fail.
+  Also, if we want to differentiate through the optimizer, √0 is not differentiable.
+  To protect against this, we add a small number, st -> st + eps2.
+  The original implementation (https://github.com/juntang-zhuang/Adabelief-Optimizer)
+  uses the square of Adam's epsilon, which we do here.
+  See also: https://github.com/juntang-zhuang/Adabelief-Optimizer/issues/61 =#
+  eps2 = o.epsilon^2
```
[the remark this answers is not in the document]
Comment (cossio) [reply src/optimise/optimisers.jl]: Ok done.

**Comment by cossio** [comment]
CI passes. Are merging this?

Also, can we tag a release with this? I could bump the version here if needed.

**Comment by mcabbott** [comment]
Does Optimisers.jl need an update to match?

**Comment by cossio** [comment]
I can do a PR to Optimisers. 

Also, +1 to releasing for each non-breaking PR. Contributors are happy to see their patches quickly available for use.

**Comment by mcabbott** [comment]
Not aware of any breaking release, I'd hope that would be well-discussed, and have a milestone. Have not paid close attention lately though. 

+1 to frequent releases.

File: src/optimise/optimisers.jl [hunk 2]
```diff
@@ -532,10 +532,26 @@ AdaBelief(η::Real, β::Tuple, state::IdDict) = AdaBelief(η, β, EPS, state)
 
 function apply!(o::AdaBelief, x, Δ)
   η, β = o.eta, o.beta
-  mt, st = get!(() -> (zero(x), zero(x)), o.state, x)::NTuple{2,typeof(x)}
+
+  mt, st, βp = get!(o.state, x) do
+      (zero(x), zero(x), Float64[β[1], β[2]])
```
Comment (ToucheSir) [inline src/optimise/optimisers.jl:537]: :vomiting_face:

I know this is what ADAM does, but it really is non-optimal. Not a blocker for this PR, but more impetus to move to Optimisers.jl ASAP.

**Review by ToucheSir — APPROVED** [review]

**Comment by darsnack** [comment]
@cossio Want to version bump then so we can make a release?

**Comment by cossio** [comment]
@darsnack Done

**Comment by ToucheSir** [comment]
Will merge this if there are no objections before the end of the day.

On release cadence, we'd be well into the hundreds on v0.12 under a 1 PR == 1 review model. Perhaps that's an argument for pushing 1.0, but to my knowledge that is (for legitimate reasons) not currently on the table. As a compromise, perhaps we could release at a somewhat regular schedule (e.g. weekly when there are new PRs, after every X PRs, after every bugfix only), and exclude non-functional changes such as docs (e.g. typo fixes) or refactoring?

**Comment by cossio** [comment]
> we'd be well into the hundreds on v0.12 under a 1 PR == 1 review model

Is there something wrong with that? There is a nice blog post by Lyndon White arguing in favor of quick releases: https://www.oxinabox.net/2019/09/28/Continuous-Delivery-For-Julia-Packages.html

On the other hand, +1 for having a v1.0. But I think we can treat this as an independent issue.

**Comment by cossio** [comment]
The FastAI failure looks unrelated to this. All other downstream tasks pass.

**Comment by cossio** [comment]
Out of curiosity, where are these downstream tests defined? I would like to know if any of them is using AdaBelief.

**Comment by ToucheSir** [comment]
> Is there something wrong with that? There is a nice blog post by `@oxinabox` arguing in favor of quick releases: https://www.oxinabox.net/2019/09/28/Continuous-Delivery-For-Julia-Packages.html

I agree with the idea of frequent releases, but much like 1.0 vs non 1.0 there's a perception issue if a repo is running 0.13.399. Too frequent updates also means more re-precompilation, which is less of a big deal locally but can add up on CI.

> Out of curiosity, where are these downstream tests defined?

https://github.com/FluxML/Flux.jl/blob/master/.github/workflows/Downstream.yml

**Pull request merged**